Add browser-like tab management hotkeys to chat overlay

angelaz1 opened the pull request:
Fixes #10148

- [x] Depends on #13314.

Adding in web browser-like keyboard shortcuts to the chat overlay:
- Closing a single chat tab with Ctrl + w 
- Pulling up the channel selector overlay with Ctrl + t to open a new tab
- Reopening the most recently closed tab that has not already been reopened Ctrl+Shift+t

Co-Authors: @jessmeng @Joseph-Ramos-CMU

jessmeng committed:
Added ctrl + w and ctrl + t keyboard shortcuts

jessmeng committed:
Fixed style changes and cleaned up ctrl + t implementation

angelaz1 committed:
Small whitespace fix

angelaz1 committed:
Adding in extra blank line

angelaz1 committed:
Removing whitespace

angelaz1 committed:
Merge pull request #3 from angelaz1/jmeng1/keyboard_shortcuts

Added ctrl + w and ctrl + t keyboard shortcuts

Joseph-Ramos-CMU committed:
Untested Ctrl+Shift+T shortcut prototype

Added a list to the ChannelManager class that tracks
which tabs I closed. Works like a stack, where it adds to the end
every time I close a tab. Then added a function that uses
this list to open the last closed channel, and added a shortcut inside of ChatOverlay,
similar to how jmeng implemented shortcuts.

Code is currently untested.

Joseph-Ramos-CMU committed:
Fixed a bug where Ctrl+Shift+t shortcut was using the wrong list

The JoinLastClosedChannel code was using the joinedChannels list instead
of the closedChannels list. Fixing this bug made the Ctrl+Shift+t
shortuct work as expected.

Joseph-Ramos-CMU committed:
Changing behvaior for channels that have already been reopened

Ctrl+Shift+t will now skip trying to reopen any channels that are
already open, and will instead attempt to open the next still-closed channel.

Joseph-Ramos-CMU committed:
Minor formatting changes

Joseph-Ramos-CMU committed:
Added Tests for the Ctrl+Shift+T shortcut

Tests have not actually been run yet.

Joseph-Ramos-CMU committed:
Additional formatting

Ran the dotnet format as well as the .\InspectCode.ps1 tools, and
fixed the style issues they found.

jessmeng committed:
Added tests for ctrl + w and ctrl + t

jessmeng committed:
Merge branch 'jmeng1/keyboard_shortcuts' of https://github.com/angelaz1/osu into jmeng1/keyboard_shortcuts

jessmeng committed:
Fixed some formatting and style

Joseph-Ramos-CMU committed:
Changing the Tests

Joseph-Ramos-CMU committed:
Temporarily removed test cases

I did this to try to figure out what was failing on pull requests

Joseph-Ramos-CMU committed:
Fixed style concerns

jessmeng committed:
Fixed formatting for tests

Joseph-Ramos-CMU committed:
Added back CI tests

angelaz1 committed:
Adding newline style fixes

Joseph-Ramos-CMU committed:
Changed the CI tests

I'm trying to make changes because something is going wrong with theGitHub CI tests.

jessmeng committed:
More style fixes

Joseph-Ramos-CMU committed:
Changed test formatting

Tests should be prefixed with [Test]. Previously my tests do not have this.

jessmeng committed:
Style fixes for whitespaces

Joseph-Ramos-CMU committed:
Toggling out CI tests

I'm still trying to track donwn why GitHub won't accept my CI.
I'm toggling tests to see what is the problem.

Joseph-Ramos-CMU committed:
Commented out all CI tests

Part of my efforts to narrow down GitHub CI issues

jessmeng committed:
Merge pull request #5 from angelaz1/jmeng1/keyboard_shortcuts

Added tests for ctrl + w and ctrl + t

Joseph-Ramos-CMU committed:
Testing first CI test

Joseph-Ramos-CMU committed:
Modified test case

I believe the tests are failing because the shortcut only works
when you select a channel on the channel dropdown. So I updated the testcases to do this.

Joseph-Ramos-CMU committed:
Minor formatting change

Still trying to fix the CI problems

Joseph-Ramos-CMU committed:
Completely removed all CI tests

THe GitHub CI is still failing for unknown reasons. I tried to
remove them all to see what the problem really is

Joseph-Ramos-CMU committed:
Added back a modified test case

Added a simple test for Ctrl+Shift+T.

Joseph-Ramos-CMU committed:
Removed all CI tests

Just trying to get a commit that passes all CI tests

angelaz1 committed:
Merge pull request #8 from angelaz1/jcramos/ctrl_shift_t

Ctrl+Shift+t Chat Shortcut

peppy commented on osu.Game.Tests/Visual/Online/TestSceneChatOverlay.cs:
```diff
@@ -208,6 +208,60 @@ public void TestChannelCloseButton()
             AddAssert("All channels closed", () => !channelManager.JoinedChannels.Any());
         }
 
+        private void pressControlW()
```
private methods should go beneath public ones.

peppy commented on osu.Game/Online/Chat/ChannelManager.cs:
```diff
@@ -33,6 +33,14 @@ public class ChannelManager : PollingComponent, IChannelPostTarget
         private readonly BindableList<Channel> joinedChannels = new BindableList<Channel>();
 
+        // Keeps a list of closed channels. More recently closed channels appear at higher indeces
+        private readonly BindableList<Channel> closedChannels = new BindableList<Channel>();
```
This should be only storing the name or ID, not the full channel object. This is going to be keeping messages, users, etc. in memory forever.

jessmeng committed:
Moved private methods beneath public ones

bdach commented on osu.Game/Online/Chat/ChannelManager.cs:
```diff
@@ -33,6 +33,14 @@ public class ChannelManager : PollingComponent, IChannelPostTarget
+        // For efficiency purposes, this constant bounds the number of closed channels we store.
+        // This number is somewhat arbitrary; future developers are free to modify it.
+        // Must be a positive number.
+        private const int closed_channels_max_size = 50;
```
We actually have an existing utility collection that does this in this repository (see `LimitedCapacityQueue`). It would have to be moved out from its current namespace first, though, so unsure if worth - and even then maybe best left for a separate pull.

bdach commented on osu.Game/Online/Chat/ChannelManager.cs:
```diff
@@ -408,13 +416,51 @@ public void LeaveChannel(Channel channel)
+            // channel in the list
+            if (joinedChannels.IndexOf(lastClosedChannel) >= 0)
+            {
+                JoinLastClosedChannel();
```
I'm not really comfortable with using recursion here, especially seeing that the worst case depth is 50 calls, and an iterative solution would work just fine.

As an aside, you could avoid this edge case whatsoever by checking if any new channel joined had been previously closed, and removing it from the list in that case. That would probably not play well with `LimitedCapacityQueue`, however, as it lacks a random access removal operation.

Joseph-Ramos-CMU replied:
Thanks for the feedback. I'll fix it to get rid of the recursion. As for avoiding the edge case, I thought about adding some deduplicate code. However, this code would have to be run whenever the user closed a tab, which could make closing a tab slow should the closedChannel list get large. I intentionally saved the checks for when the user actually used the Ctrl+Shift+t shortcut, this way the new feature would have minimal impact on the performance of any other part of the code.

bdach replied:
I mean I don't think it would be *that* much worse than the current code. Let's denote the number of previously-closed channels with `n`, and the number of currently-joined channels with `m`.

Current code:

* An item is added to `closedChannels` on every channel close, which is an amortised `O(1)` operation - amortised because the list has to be expanded in `O(n)` time every now and again, but as far as I know .NET uses the proper exponential reallocation scheme to achieve amortised `O(1)`. (As an aside, `closedhannels` is not currently checked for duplicates when added to, which seems unintended.)
* Each attempt to restore a channel is an `O(m)` operation (worst case for the `IndexOf()` to check if the channel we're trying to restore is open), and there can be up to `n` attempts, yielding a total of `O(n * m)`.

What I'm proposing:

* Adding to `closedChannels` would be the same as in the current code. (Preferably with duplicate checks.)
* When joining a channel, we'd have to check closed channels to make sure we clean them up, which is an `O(n)` operation. Joining `m` channels would therefore be `O(n * m)` again (but in `m` different operations instead of one).
* Restoring the last closed channel would be `O(1)` again, as the only thing to be done is to rejoin the last closed-but-not-reopened channel from `closedChannels`.

I guess the argument could be made that joining channels is a more frequent operation than restoring the last closed one, but I'm not sure `m` and `n` will ever get large enough where that will be a problem. That said I don't insist on this solution - I'm fine with just removing the recursion.

bdach commented on osu.Game/Overlays/ChatOverlay.cs:
```diff
@@ -357,6 +357,32 @@ protected override bool OnKeyDown(KeyDownEvent e)
+                {
+                    switch (e.Key)
+                    {
+                        case Key.T:
```
We have a framework-side mechanism called platform actions that might be good to use here, as tab operations are pretty well defined UI interactions overall. I can prepare a pull adding those.

peppy replied:
@bdach Shall we get this done framework side and try and merge this PR?

Should probably be added for the "close tab" key below as well (ctrl-w on windows; cmd-q on macOS).

bdach replied:
I already have added those actions, just never got round to updating this pull with them as it kept falling down the backlog.

peppy replied:
I've made the required changes here. @bdach can you give this a once-over whenever you have a chance?

bdach reviewed (changes requested):
One big issue (the recursion usage); rest is mostly code quality stuff.

Joseph-Ramos-CMU committed:
Changed internal closedChannel list to store channels names instead of channel objects

This is for efficiency purposes; it's more efficient to just store
the names than store the whole object.

Joseph-Ramos-CMU committed:
Reworked reopening last tab to no longer use recursion

A reviewer of the pull request was concerned about recursion. I changed
the code to be iterative.

Joseph-Ramos-CMU committed:
Formatting fixes

Fixed some formatting issues to comply with InspectCode.

bdach commented:
On reading this diff again I noticed that the <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>T</kbd> hotkey is not tested. It would be nice to have coverage for this as well.

angelaz1 committed:
Making style changes + supports reopening PM chats

angelaz1 committed:
Merge branch 'keyboard_shortcuts' into jcramos/ctrl_shift_t

angelaz1 committed:
Minor test style fixes

angelaz1 committed:
CI Style Fixes

angelaz1 committed:
Whitespace fixes

Joseph-Ramos-CMU committed:
Merge pull request #10 from angelaz1/jcramos/ctrl_shift_t

Fixes based on reviewer pull request feedback

Joseph-Ramos-CMU commented:
I believed we've addressed all the critical changes you requested. As for test coverage for Ctrl+Shift+t, we tried adding tests but were having a lot of issues with Appveyor. The commit history should show some of the tests we tried. Pretty much followed the same format as the Ctrl+W and Ctrl+T tests, but it wasn't working for some reason. It might be because the Ctrl+Shift+t shortcut only works if the user has the chat tabs selected, but even after updating the tests to account for this we still had issues.

Let us know if there are any other must-do fixes.

bdach commented on osu.Game/Online/Chat/ChannelManager.cs:
```diff
@@ -408,13 +419,72 @@ public void LeaveChannel(Channel channel)
+                    }
+                    else
+                    {
+                        // Try to get User to open PM chat
+                        var req = new GetUserRequest(lastClosedChannelId);
+                        req.Success += user => CurrentChannel.Value = JoinChannel(new Channel(user));
+                        api.Queue(req);
```
This will unnecessarily re-fetch the user every time. The proper solution for this is to use `UserLookupCache` which will 100% have the user locally fetched already.

angelaz1 committed:
Review changes + added tests

angelaz1 committed:
Style fixes

angelaz1 committed:
Merge pull request #11 from angelaz1/jcramos/ctrl_shift_t

Made review changes + added tests

peppy committed:
Merge branch 'master' into keyboard_shortcuts

peppy committed:
Merge branch 'master' into keyboard_shortcuts

peppy committed:
Mark fields `readonly`

peppy committed:
Replace second usage of new function

peppy committed:
Use new platform actions instead of hardcoded keys

peppy committed:
Update tests to check for correct platform action keys

peppy committed:
Merge branch 'test-platform-actions' into keyboard_shortcuts

bdach commented:
There's still some weirdness left in this PR that I hadn't seen or had time to address properly yet. Going to try and clean this up more myself now.

bdach committed:
Invert condition to reduce nesting

bdach committed:
Use standard list instead of bindable list

No reason to use a bindable list there, as `CollectionChanged` was never
subscribed to.

bdach committed:
Merge branch 'master' into keyboard_shortcuts

bdach committed:
Rename method to be less misleading

Would rather avoid variations of Equals/Equal/IsEqual. There's not
really much equality involved as the types are different.

bdach committed:
General comment cleanups

bdach committed:
Rename test steps

bdach commented:
Have pushed my changes, mostly cosmetic. I did spot one more issue, namely that when the channel selector overlay is open, the hotkeys stop working - I think it's because the channel selector is taking  focus away from the chat overlay, so I'm not sure that's easily fixable. Other than that seems passable.

@peppy feel free take a look at the latest changes if you're so inclined :)
